pyterrier/model.py

```diff
@@ -85,80 +85,134 @@ def query_columns(df : pd.DataFrame, qid=True) -> Sequence[str]:
-            prev_col = next_col
-        else:
-             break
-    if keep_original:
-        inp['query'] = inp['query_0']
-    return inp
+    def per_element(i: IterDictRecord) -> IterDictRecord:
+        cols = i.keys()
+        if base_column not in cols:
+            raise KeyError(f"Expected a {base_column} column, but found {list(cols)}")
+        prev_col = base_column
+        rename_cols = {}
+        for query_idx in itertools.count():
+            next_col = f"{base_column}_{query_idx}"
+            if prev_col in cols:
+                rename_cols[prev_col] = next_col
+                prev_col = next_col
+            else:
+                break
+
+        renamed = {}
+        for k, v in i.items():
+            if k in rename_cols:
+                renamed[rename_cols[k]] = v
+            else:
+                renamed[k] = v
+
+        if keep_original:
+            renamed[base_column] = renamed[f"{base_column}_0"]
+
+        return renamed
+
+    if isinstance(inp, dict):
+        return per_element(inp)
+    return [*map(per_element, inp)]
 
-
-def pop_queries(df: pd.DataFrame, *, inplace: bool = False) -> pd.DataFrame:
+def push_queries(df: pd.DataFrame, *, keep_original: bool = False, inplace: bool = False) -> pd.DataFrame:
     """
-        Changes a dataframe such that the "query_0" column becomes "query_1", and any
-        "query_1" columns becames "query_0" etc. In effect, does the opposite of push_queries().
-        The current "query" column is dropped.
+        Changes a dataframe such that the "query" column becomes "query_0", and any
+        "query_0" columns becames "query_1" etc.
 
         Arguments:
             df: Dataframe with a "query" column
+            keep_original: if True, the query column is also left unchanged. Useful for client code. 
+                Defaults to False.
             inplace: if False, a copy of the dataframe is returned. If True, changes are made to the
                 supplied dataframe. Defaults to False. 
     """
+    assert not inplace, "push_queries no longer supports inplace"
+    return push_columns(df, keep_original=keep_original, inplace=inplace, base_column="query")
+
+
+@overload
+def push_queries_dict(inp: IterDictRecord, keep_original: bool = False, inplace: bool = False) -> IterDictRecord: ...
+@overload
+def push_queries_dict(inp: IterDict, keep_original: bool = False, inplace: bool = False) -> IterDict: ...
+
+def push_queries_dict(inp: Union[IterDictRecord,IterDict], keep_original: bool = False, inplace: bool = False) -> Union[IterDictRecord,IterDict]:
+    """
+    Works like ``push_queries`` but over a dict/iterdict instead of a dataframe.
+    """
+    assert not inplace, "push_queries_dict does not support inplace"
+    return push_columns_dict(inp, keep_original=keep_original, base_column="query")
+
+
+def pop_columns(df: pd.DataFrame, base_column="query") -> pd.DataFrame:
```

Comment:
Are we missing a pop for the dict case? This also isn't a regression, just surprising.

Reply:
yes